src/Build/Instance/ProjectPropertyInstance.cs

```diff
@@ -84,8 +86,21 @@ public string EvaluatedValue
         /// Setter assumes caller has protected global properties, if necessary.
         /// </summary>
         [DebuggerBrowsable(DebuggerBrowsableState.Never)]
-        string IProperty.EvaluatedValueEscaped => _escapedValue;
-
+        string IProperty.EvaluatedValueEscaped
+        {
+            get
+            {
+                if (this is EnvironmentDerivedProjectPropertyInstance envProperty && envProperty.loggingContext?.IsValid == true && !envProperty._loggedEnvProperty)
+                {
+                    EnvironmentVariableReadEventArgs args = new(Name, _escapedValue);
+                    args.BuildEventContext = envProperty.loggingContext.BuildEventContext;
+                    envProperty.loggingContext.LogBuildEvent(args);
+                    envProperty._loggedEnvProperty = true;
```

Comment:
[Q after a quick first pass]

This is the only place I see where `EnvironmentDerivedProjectPropertyInstance` is marked as "logged", is it true that it would only ever be considered logged whenever `EvaluatedValueEscaped` is called?

Reply:
Is this the only point that the env var would get added to the binlog?

Reply:
Correct on only considered logged when EvaluatedValueEscaped is called. That's the only way an EnvironmentDerivedProjectPropertyInstance's environment-derived property could get added to the binlog, but I also added EnvironmentDerivedProjectProperty, and that's another place environment-derived properties could get added.

Reply:
good catch, I think there's a loophole where other properties on `ProjectPropertyInstance` are called that effectively constitute a "read", but we don't log those.

Reply:
@Forgind let's revisit this - I think this might be potentially the cause of our problems